### ISSUE TITLE
Cannot use non-int default value with to_yojson

### ISSUE THREAD

**Issue by alexozer** [ocaml-ppx/ppx_deriving_yojson#79]
Applying default value to record fields that aren't of type int seem to fail with a type error when using `[@@deriving to_yojson]`:

This seems to work:

```ocaml
type thing = {
  n : int [@default 0];
  str : string [@default "sdfasdf"];
  lst : int list;
}
[@@deriving of_yojson]
```

But when using to_yojson or yojson instead of of_yojson:

```ocaml
type thing = {
  n : int [@default 0];
  str : string [@default "sdfasdf"];
  lst : int list;
}
[@@deriving to_yojson]
```

I get:

```
Error: This expression has type string but an expression was expected of type
         int
```

Eliminating the default on string works with to_yojson and yojson, however:

```
type thing = {
  n : int [@default 0];
  str : string;
  lst : int list;
}
[@@deriving to_yojson]
```

**Comment by gasche** [comment]
Which version of OCaml and ppx_deriving_yojson are you using? I cannot reproduce the problem under 4.06.1 or 4.07.0. The generated code is as follows (`ocamlfind ocamlc -package ppx_deriving_yojson -dsource -c test.ml`):

```
type thing =
  {
  n: int [@default 0];
  str: string [@default "sdfasdf"];
  lst: int list }[@@deriving to_yojson]
let rec (thing_to_yojson : thing -> Yojson.Safe.json) =
  ((let open! Ppx_deriving_yojson_runtime in
      fun x ->
        let fields = [] in
        let fields =
          ("lst",
            ((fun x ->
                `List
                  (List.map (fun (x : Ppx_deriving_runtime.int) -> `Int x) x))
               x.lst))
          :: fields in
        let fields =
          if x.str = "sdfasdf"
          then fields
          else
            ("str",
              (((fun (x : Ppx_deriving_runtime.string) -> `String x)) x.str))
            :: fields in
        let fields =
          if x.n = 0
          then fields
          else ("n", (((fun (x : Ppx_deriving_runtime.int) -> `Int x)) x.n))
            :: fields in
        `Assoc fields)
  [@ocaml.warning "-A"])
```

In particular, `"sdfasdf"` is correctly tested against the `x.str` type, which has type `string` as expected.

**Comment by Chimrod** [comment]
i can reproduce it if I insert `open Base` before the definition. This give the following code :

```ocaml
open Base
type thing =
  {
  n: int [@default 0];
  str: string [@default "sdfasdf"];
  lst: int list }[@@deriving to_yojson]
let rec (thing_to_yojson : thing -> Yojson.Safe.json) =
  ((let open! Ppx_deriving_yojson_runtime in
      fun x ->
        let fields = [] in
        let fields =
          ("lst",
            ((fun x ->
                `List
                  (List.map (fun (x : Ppx_deriving_runtime.int) -> `Int x) x))
               x.lst))
          :: fields in
        let fields =
          if x.str = "sdfasdf"
          then fields
          else
            ("str",
              (((fun (x : Ppx_deriving_runtime.string) -> `String x)) x.str))
            :: fields in
        let fields =
          if x.n = 0
          then fields
          else ("n", (((fun (x : Ppx_deriving_runtime.int) -> `Int x)) x.n))
            :: fields in
        `Assoc fields)
  [@ocaml.warning "-A"])[@@ocaml.warning "-39"]
let _ = thing_to_yojson
File "test.ml", line 2, characters 0-118:
Error: This expression has type Base.string = string
       but an expression was expected of type int
```

**Comment by gasche** [comment]
So I guess that the issue comes from Base redefining `(=)` to work on integers only. This means that the code that emits `(=)` should explicitly use it from the Ppx_deriving runtime.

**Comment by lezed1** [comment]
It looks like `Base` shadows `( = )` to be `int -> int -> bool` (making it non-polymorphic). This would cause the [generated code](https://github.com/ocaml-ppx/ppx_deriving_yojson/blob/cadb5f653d4c2a94bc9e601c02dac75698cd5b97/src/ppx_deriving_yojson.cppo.ml#L260) to only work on `int`.

Here are the relevant parts of `Base`:
[Defining `( = )`](https://github.com/janestreet/base/blob/c81240c038e434c8fa4e576bc941d0f8b1fa7fb8/src/import0.ml#L156)
[Including `( = )`:](https://github.com/janestreet/base/blob/c81240c038e434c8fa4e576bc941d0f8b1fa7fb8/src/base.ml#L387)

The simplest thing I can think of is re-shadowing comparison operators to be [polymorphic](https://github.com/janestreet/base/blob/c81240c038e434c8fa4e576bc941d0f8b1fa7fb8/src/poly0.mli): 
```ocaml
open Base
open Poly
```

A cool long term idea could be specifying a equals function. This could let users drop floats if they are similar, etc. A possible syntax could be this:
```ocaml
type t =
  { f : float [@default ~equals:some_function 1.23]
  } [@@deriving to_yojson]
```

**Comment by diml** [comment]
FTR, for Jane Street ppx rewriters we got into the habit of systematically adding runtime libraries and generating code that references the runtime library. This make the rewriters more robust as they make much less assumption about the environment.

**Comment by gasche** [comment]
Yes, that's also the general ppx_deriving approach, but I guess that this particular function slipped up.

I wonder if we could have some tooling support for these hygiene-breakage detection. One idea would be to try to statically analyze the preprocessor code (all identifiers accessed by the code must be either local or to a list of explicitly specified modules), or to use lightweight annotations in the code (possibly only in a debugging mode) that let us distinguish user-provided from processor-provided code in the expansion result, and perform a check there.

**Comment by diml** [comment]
A static analyzer would be nice. A simpler version would be to have tests compiled in an empty initial environment.

**Comment by vphantom** [comment]
Wouldn't the simplest short-term solution be to shadow with an explicit `let (=) = Stdlib.(=)` in `Ppx_deriving_yojson_runtime` to benefit the generated code?

It seems so trivial to add, that I bet I'm missing some obvious reason why that wouldn't be appropriate. :thinking:

**Comment by whitequark** [comment]
> It seems so trivial to add, that I bet I'm missing some obvious reason why that wouldn't be appropriate. thinking

No, it's just an oversight:

> Yes, that's also the general ppx_deriving approach, but I guess that this particular function slipped up.

**Comment by vphantom** [comment]
Should I make this my very first OCaml PR? :wink:

**Comment by cemerick** [comment]
FWIW, here's a workaround that will make the derived code happy, but allow you to keep your preferred monomorphic `=`:

```ocaml
module Workaround = struct
  let (=) = Stdlib.(=)
  module T = struct
    type foo = { bar : bool [@default true]}
               [@@deriving yojson]
  end
end
include Workaround.T
```

Just drop your types in `T`.

**Comment by gasche** [comment]
@cemerick, @vphantom, would one of you like to write a PR to fix the bug upstream?

**Comment by vphantom** [comment]
I have been away from OCaml for… wow over a year now (can't wait to come back!) so I'm a bit rusty, but I can definitely give it a shot. :smiley:

**Comment by cemerick** [comment]
I can take a shot if @vphantom decides to pass.

**Comment by vphantom** [comment]
Hey @cemerick since I'm a bit out of touch I suggest you look into it then.

**Comment by cemerick** [comment]
Darn, it looks like this was already fixed with #126 (of _course_ I notice only after fixing it in a slightly different way on my fork! 🙃).

I can't seem to build or install the library at the relevant commit, but I suspect that's a problem with my current switch or something.
